### PR TITLE
Promote qa to prod

### PR DESCRIPTION
**Pull request by peintnermax**
These features / fixes were implemented with this PR

- password change required is handled now. after a user password is checked, and if change is required, a user will be redirected to `/password/change`
- an additional check for force MFA is done, after checking a user password. If either forceMFA or forceLocalMFA is enforced, a user is prompted to setup MFA after password verification
- a password reset link now includes the authRequest if the flow was initiated with it 
- OTP links redirect to the same host where the flow was initiated and include the authRequest
- if phone / email login is disabled, user results are removed where the input equals the disabled method
- If passkeys / password is disabled, throw an error.

### CODE REVIEW

File: apps/login/src/components/login-otp.tsx [hunk 1]
```diff
@@ -76,7 +78,18 @@ export function LoginOTP({
...
     if (method === "email") {
       challenges = create(RequestChallengesSchema, {
-        otpEmail: { deliveryType: { case: "sendCode", value: {} } },
+        otpEmail: {
+          deliveryType: {
+            case: "sendCode",
+            value: host
+              ? {
+                  urlTemplate:
+                    `${host.includes("localhost") ? "http://" : "https://"}${host}/otp/method=${method}?code={{.Code}}&userId={{.UserID}}&sessionId={{.SessionID}}&organization={{.OrgID}}` +
```
Comment (eliobischof) [inline apps/login/src/components/login-otp.tsx:87]: Why not using the origin?

```suggestion
                    `${origin}/otp/method=${method}?code={{.Code}}&userId={{.UserID}}&sessionId={{.SessionID}}&organization={{.OrgID}}` +
```

File: apps/login/src/lib/server/idp.ts [hunk 2]
```diff
@@ -9,11 +10,17 @@ export type StartIDPFlowCommand = {
...
+    return { error: "Could not get host" };
+  }
+
   return startIdentityProviderFlow({
     idpId: command.idpId,
     urls: {
-      successUrl: command.successUrl,
-      failureUrl: command.failureUrl,
+      successUrl: `${host.includes("localhost") ? "http://" : "https://"}${host}${command.successUrl}`,
+      failureUrl: `${host.includes("localhost") ? "http://" : "https://"}${host}${command.failureUrl}`,
```
Comment (eliobischof) [inline apps/login/src/lib/server/idp.ts:23]: use origin?

**Review by eliobischof — CHANGES_REQUESTED** [review]

File: apps/login/src/components/login-otp.tsx (hunk 1 above)
Comment (peintnermax) [reply apps/login/src/components/login-otp.tsx:87]: We can not get the origin outside of a route handler, so we use the host header

Comment (eliobischof) [reply apps/login/src/components/login-otp.tsx:87]: We can't read the origin header? Are you sure about that?

Comment (peintnermax) [reply apps/login/src/components/login-otp.tsx:87]: The problem is that the origin header is not sent everywhere

File: apps/login/src/lib/server/password.ts [hunk 3]
```diff
@@ -192,24 +213,14 @@ export async function sendPassword(command: UpdateSessionCommand) {
...
-      params.append(
-        "organization",
-        command.organization ?? session.factors?.user?.organizationId,
-      );
-    }
-
-    return { redirect: `/password/change?` + params };
-  } else if (command.forceMfa && !availableSecondFactors.length) {
+  }
+  // TODO: check if handling of userstate INITIAL is needed
```
Comment (eliobischof) [inline apps/login/src/lib/server/password.ts:217]: I would say no, as it does not exist on v2 api users

File: apps/login/src/lib/server/loginname.ts [hunk 4]
```diff
@@ -35,6 +36,15 @@ export async function sendLoginname(command: SendLoginnameCommand) {
 
   const loginSettings = await getLoginSettings(command.organization);
 
+  const potentialUsers = users.result.filter((u) => {
+    const human = u.type.case === "human" ? u.type.value : undefined;
+    return loginSettings?.disableLoginWithEmail
+      ? human?.email?.isVerified && human?.email?.email !== command.loginName
+      : loginSettings?.disableLoginWithPhone
+        ? human?.phone?.isVerified && human?.phone?.phone !== command.loginName
```
Comment (eliobischof) [inline apps/login/src/lib/server/loginname.ts:44]: should this not rather match? like human?.email?.isVerified && human?.email?.email === command.loginName

`===` istead of `!==` ?

File: apps/login/src/lib/server/password.ts [hunk 5]
```diff
@@ -148,6 +148,27 @@ export async function sendPassword(command: UpdateSessionCommand) {
...
+
+    if (command.organization || session.factors?.user?.organizationId) {
+      params.append("organization", session.factors?.user?.organizationId);
+    }
+
+    if (command.authRequestId) {
+      params.append("authRequestId", command.authRequestId);
+    }
+
+    return { redirect: "/password/change?" + params };
```
[the remark this answers is not in the document]
Comment (peintnermax) [reply apps/login/src/lib/server/password.ts:169]: exactly

File: apps/login/src/lib/server/password.ts (hunk 3 above)
Comment (peintnermax) [reply apps/login/src/lib/server/password.ts:217]: agree. I keep the comments for the moment though as they help to identify the use cases where the user state is critical

File: apps/login/src/lib/server/loginname.ts (hunk 4 above)
Comment (peintnermax) [reply apps/login/src/lib/server/loginname.ts:44]: The idea is that we exclude users from the query using the loginname query. if email is not allowed, we only allow users where email is not on the resource thus !==

**Review by eliobischof — APPROVED** [review]

**Pull request merged**